src/sctools/test/test_count.py

```diff
@@ -1,44 +1,578 @@
+         unit, in correspondence with a randomly generated count matrix. As such, the primary alignment records
+         alone are expected to reproduce the randomly generated count matrix.
+
+      2) duplicate alignments: randomly picked from primary alignments though given a new query name. These
+         record correspond to PCR and optical duplicates. `count.from_sorted_tagged_bam` must ignore such reads.
+
+      3) incomplete alignments: these records miss at least one necessary tag. `count.from_sorted_tagged_bam` must
+         ignore such reads.
+
+      4) multi-gene alignments: these records have the same tags and query name, though, at least two such
+         records exist that point to different genes. `count.from_sorted_tagged_bam` must ignore such reads.
+
+    Parameters
+    ----------
+    num_cells : int
+        number of real cells
+    max-genes : int
+        maximum number of genes to use to generate synthetic counts
+    annotation_file : str
+        annotation GTF file to look up gene names from
+    gene_expression_rate : float
+        poisson rate at which each gene is expressed
+    rng_seed : int
+        random number generator seed
+
+    Methods
+    -------
+    generate_synthetic_counts_and_alignment_tags
+        generates synthetic test data as described above and returns an instance of SyntheticDataBundle
+    generate_synthetic_bam_and_counts_matrix
+        generates synthetic test data and writes them to disk
+
+    See Also
+    --------
+    count.from_sorted_tagged_bam
+    """
+
+    cell_barcode_tag = 'CB'
+    molecule_barcode_tag = 'UB'
+    gene_name_tag = 'GE'
+    output_prefix = 'synthetic_'
+    bam_output_filename = output_prefix + 'records.bam'
+    count_matrix_output_filename = output_prefix + 'count_matrix.npy'
+    row_index_output_filename = output_prefix + '_row_index.npy'
+    col_index_output_filename = output_prefix + '_col_index.npy'
+
+    def __init__(self,
+                 num_cells: int,
+                 max_genes: int,
+                 annotation_file: str,
+                 gene_expression_rate: float,
+                 rng_seed: int = 777):
+        self.num_cells = num_cells
+        self.gene_expression_rate = gene_expression_rate
+
+        # initialize the random number generator
+        self.rng: np.random.RandomState = np.random.RandomState(seed=rng_seed)
+
+        # generate gene names
+        self.all_gene_names = self._load_gene_names(annotation_file)
+        self.max_genes = max_genes
+        self.num_genes = len(self.all_gene_names)
+        self.to_be_used_gene_indices: List[int] = self.rng.choice(
+            np.arange(0, self.num_genes, dtype=np.int), size=self.max_genes, replace=False).tolist()
+        self.to_be_used_gene_names = [self.all_gene_names[j] for j in self.to_be_used_gene_indices]
+
+    def generate_synthetic_counts_and_alignment_tags(self,
```

Comment:
I like parameter lists to go on the next line to save on leading white space. Also because it always works--in cases where you have really long function + parameter names the style used here can overrun your line limit.

Reply:
Done.